sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -0,0 +1,186 @@
+     * parser. If it is an UnresolvedAttribute, we look it up in the catalog and return it. If it
+     * is an Alias, we resolve the child and return an Alias with the same name.
+     */
+    e match {
+      case u: UnresolvedAttribute =>
+        getVariableReference(u.nameParts)
+      case a: Alias =>
+        Alias(resolveVariable(a.child), a.name)()
+
+      case other =>
+        throw SparkException.internalError(
+          "Unexpected variable expression in ParametrizedQuery: " + other)
+    }
+  }
+
+  def resolveArguments(expressions: Seq[Expression]): Seq[Expression] = {
+    expressions.map { exp =>
+      if (exp.resolved) {
+        exp
+      } else {
+        resolveVariable(exp)
+      }
+    }
+  }
+
+  def extractQueryString(either: Either[String, UnresolvedAttribute]): String = {
+    either match {
+      case Left(v) => v
+      case Right(u) =>
+        val varReference = getVariableReference(u.nameParts)
+
+        if (!varReference.dataType.sameType(StringType)) {
+          throw new AnalysisException(
+            errorClass = "INVALID_VARIABLE_TYPE_FOR_QUERY_EXECUTE_IMMEDIATE",
+            messageParameters = Map("varType" -> varReference.dataType.simpleString))
+        }
+
+        // Call eval with null value passed instead of a row.
+        // This is ok as this is variable and invoking eval should
+        // be independent of row value.
+        varReference.eval(null).toString
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan =
+    plan.resolveOperatorsWithPruning(_.containsPattern(EXECUTE_IMMEDIATE), ruleId) {
+      case ExecuteImmediateQuery(expressions, query, targetVariablesOpt, parser) =>
+        val queryString = extractQueryString(query)
+        val plan = parseStatement(parser, queryString, targetVariablesOpt)
+
+        val posNodes = plan.collect { case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: PosParameter => n })
+        }.flatten
+        val namedNodes = plan.collect { case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: NamedParameter => n })
+        }.flatten
+
+        val queryPlan = if (expressions.isEmpty || (posNodes.isEmpty && namedNodes.isEmpty)) {
+          plan
+        } else if (posNodes.nonEmpty && namedNodes.nonEmpty) {
+          throw new AnalysisException(
+            errorClass = "INVALID_QUERY_BOTH_POSITIONAL_AND_NAMED_PARAMETERS_PRESENT",
+            messageParameters = Map.empty)
+        } else {
+          if (posNodes.nonEmpty) {
+            PosParameterizedQuery(
+              plan,
+              // We need to resolve arguments before Resolution batch to make sure
+              // that some rule does not accidently resolve our parameters.
+              // We do not want this as they can resolve some unsupported parameters
+              resolveArguments(expressions))
+          } else {
+            val namedExpressions = expressions.collect { case (e: NamedExpression) => e }
```

Comment:
shall we fail or ignore un-named parameters? cc @srielau

Reply:
we are allowing empty keys tho.
<img width="783" alt="image" src="https://github.com/apache/spark/assets/150366084/2b7271b7-1c25-4f1f-9320-15fa0f5568e9">